SimPEG/utils/model_utils.py

```diff
@@ -195,3 +200,128 @@ def depth_weighting(
+        row should contain the coordinates of a single location point in the
+        following order: _x_, _y_, _z_ (for 3D meshes) or _x_, _z_ (for 2D
+        meshes).
+        The coordinate of the reference location, usually the receiver locations
+    active_cells : (mesh.n_cells) numpy.ndarray of bool, optional
+        Index vector for the active cells on the mesh.
+        If ``None``, every cell will be assumed to be active.
+    exponent : float, optional
+        Exponent parameter for distance weighting.
+        The exponent should match the natural decay power of the potential
+        field. For example, for gravity acceleration, set it to 2; for magnetic
+        fields, to 3.
+    threshold : float or None, optional
+        Threshold parameters used in the distance weighting.
+        If ``None``, it will be set to half of the smallest cell width.
+    engine: str, 'numpy' or 'numba': pick between a numpy vectorized computation (memory intensive) or parallelized
+        numba implementation. Default to 'numba'.
+
+    Returns
+    -------
+    (n_active) numpy.ndarray
+        Normalized distance weights for the mesh at every active cell as
+        a 1d-array.
+    """
+
+    active_cells = (
+        np.ones(mesh.n_cells, dtype=bool) if active_cells is None else active_cells
+    )
+
+    # Default threshold value
+    if threshold is None:
+        threshold = 0.5 * mesh.h_gridded.min()
+
+    reference_locs = np.asarray(reference_locs)
+
+    cell_centers = mesh.cell_centers[active_cells]
+    cell_volumes = mesh.cell_volumes[active_cells]
+
+    # address 1D case
+    if mesh.dim == 1:
+        cell_centers = cell_centers.reshape(-1, 1)
+        reference_locs = reference_locs.reshape(-1, 1)
+
+    if engine == "numba":
+
+        @njit(parallel=True)
+        def distance_weighting_numba(
+            cell_centers: np.ndarray,
+            cell_volumes: np.ndarray,
+            reference_locs: np.ndarray,
+            threshold: float,
+            exponent: float = 2.0,
+        ):
+            distance_weights = np.zeros(len(cell_centers))
+            for _, rl in enumerate(reference_locs):
+                dst_wgt = (
+                    np.sqrt(((cell_centers - rl) ** 2).sum(axis=1)) + threshold
+                ) ** exponent
+                dst_wgt = (cell_volumes / dst_wgt) ** 2
+                distance_weights += dst_wgt
+
+            return distance_weights
+
+        dist_weights = distance_weighting_numba(
+            cell_centers,
+            cell_volumes,
+            reference_locs,
+            exponent=exponent,
+            threshold=threshold,
+        )
+
+    elif engine == "numpy":
+        n, d = cell_centers.shape
+        t, d1 = reference_locs.shape
+        if not d == d1:
+            raise Exception("vectors must have same number of columns")
+
+        # vectorized distance calculations
+        distance = (
```

Comment:
i suspect there is a more memory-efficient way to do this-- you might try one of the scipy functions https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.distance.cdist.html#scipy.spatial.distance.cdist